qiskit/circuit/quantumcircuit.py

```diff
@@ -4348,38 +4351,57 @@ def assign_parameters(  # pylint: disable=missing-raises-doc
 
         if isinstance(parameters, collections.abc.Mapping):
             raw_mapping = parameters if flat_input else self._unroll_param_dict(parameters)
-            our_parameters = self._data.unsorted_parameters()
-            if strict and (extras := raw_mapping.keys() - our_parameters):
+            if strict and (
+                extras := [
+                    parameter for parameter in raw_mapping if not self.has_parameter(parameter)
+                ]
+            ):
                 raise CircuitError(
                     f"Cannot bind parameters ({', '.join(str(x) for x in extras)}) not present in"
                     " the circuit."
                 )
-            parameter_binds = _ParameterBindsDict(raw_mapping, our_parameters)
-            target._data.assign_parameters_mapping(parameter_binds)
+
+            def create_mapping_view():
+                return raw_mapping
+
+            target._data.assign_parameters_mapping(raw_mapping)
         else:
-            parameter_binds = _ParameterBindsSequence(target._data.parameters, parameters)
+            # This should be cause a cache retrieval, since we warmed the cache.  We need to keep
```

Comment:
```suggestion
            # This should cause a cache retrieval, since we warmed the cache.  We need to keep
```

Reply:
Also, to make sure I understand, we want to save a reference to the initial parameters since `target._data.parameters` will be updated as we go? I think the code comment here might benefit from some more clarification. I'm not sure I understand the bit about us warming the cache. The `mapping_view` is what is cached to my knowledge, not the `initial_parameters`?

Reply:
I expanded all the comments in fcc0c43999.

The general idea is that if there's calibrations, Python space needs to construct a `{parameter: bound_value}` mapping if we're called in the sequence form.  However, for large numbers of parameters, that's quite expensive to construct, and the vast vast vast majority of circuits don't have parametrised calibrations, so we don't want to do it.  To construct the mapping lazily, we need to keep hold of what the parameters were initially, which is derived from the Rust-space `ParameterTable`.  Later, the `mapping_view` is that `{parameter: bound_value}` map, but it's lazily constructed in `map_calibrations` on the first access.

Reply:
Excellent, the improvements you've made to the code comments make it much clearer. Thanks!